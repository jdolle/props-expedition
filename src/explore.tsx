/**
 * The core of the package. Explore.
 */

import * as React from 'react'

export type TransformFn<Props, ExploreProps, Key extends keyof ExploreProps> =
  (nextProps?: Props, props?: Props | {}, value?: ExploreProps[Key]) => ExploreProps[Key] | Promise<ExploreProps[Key]>

export type ExploreConfig<Props, ExploreProps> = {
  [Key in keyof ExploreProps]: TransformFn<Props, ExploreProps, Key>;
}

type ExploreState<ExploreProps> = {
  [Key in keyof ExploreProps]?: ExploreProps[Key]
}

/**
 * A higher order component that brings your props on a trip.
 */
export function explore<
  Props extends { [s: string]: any },
  ExploreProps extends { [s: string]: any }
>(transform: ExploreConfig<Props, ExploreProps>) {

  return (WrappedComponent: React.StatelessComponent<Props> | React.ComponentClass<Props>) => {
    type State = ExploreState<ExploreProps>

    return class Explorer extends React.PureComponent<Props, State> {
      public static displayName = `Explored(${WrappedComponent.displayName})`

      constructor(props: Props) {
        super(props)
        const initialState: ExploreProps = Object.keys(transform).reduce(
          (sumState: State, key: string) => {
            const value = transform[key](props, {}, undefined)

            if (value instanceof Promise) {
              this.resolvePromiseToKey(value, key)

            } else {
              sumState[key] = value
            }

            return sumState
          },
          {},
        ) as ExploreProps
        this.state = initialState

        // NOTE( JMD): Need to somehow handle errors from async calls
        // const errors = {}
        // NOTE (JMD): Spread on Generics is bugged https://github.com/Microsoft/TypeScript/issues/10727
        // tslint:disable-next-line prefer-object-spread
        // const fullProps: Props & State = Object.assign({}, this.props, this.state, errors)

        // if (WrappedComponent instanceof React.PureComponent || WrappedComponent instanceof React.Component) {
        //   const errors = {}
        //   WrappedComponent.render = WrappedComponent.render.bind(WrappedComponent, fullProps, errors)
        // }
      }

      public componentWillReceiveProps(nextProps: Readonly<Props>): void {
        Object.keys(transform).forEach(async (key: string) => {
          const value = transform[key](nextProps, this.props, this.state[key])

          if (value instanceof Promise) {
            this.resolvePromiseToKey(value, key)

          } else {
            this.setState({[key]: value})
          }
        })
      }

      public render(): JSX.Element {
        // NOTE (JMD): Spread on Generics is bugged https://github.com/Microsoft/TypeScript/issues/10727
        // tslint:disable-next-line prefer-object-spread
        const fullProps: Props & State = Object.assign({}, this.props, this.state)

        return (
          <WrappedComponent {...fullProps}/>
        )
      }

      // tslint:disable-next-line
      // TODO (JMD): Make private. Figure out https://github.com/Microsoft/TypeScript/issues/17293
      public resolvePromiseToKey<T>(promise: Promise<T>, key: string) {
        promise
          .then((data) => {
            this.setState({
              [key]: data,
            })
          })
          .catch(() => {
            this.setState({
              [key]: undefined, // Async value got rejected. Set to undefined.
            })
          })
      }
    }
  }
}
