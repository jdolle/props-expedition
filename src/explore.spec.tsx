/**
 * Test the Explore HOC
 */

import {shallow} from 'enzyme'
import * as React from 'react'

import {explore} from 'explore'

// A simple component used to check if the props are being passed correctly
function BasicComponent<Props extends { [s: string]: any}>(props: Props): JSX.Element {
  let out: string
  try {
    out = JSON.stringify(props)
  } catch (e) {
    out = ''
  }

  return (
    <div>{out}</div>
  )
}

describe('explore()', () => {
  describe('constructor', () => {
    it('forwards nothing for {}', () => {
      const ExploredComponent = explore({})(BasicComponent)

      expect(shallow(<ExploredComponent/>)).toMatchSnapshot()
    })

    it('forwards basic props', () => {
      type ExampleComponentProps = {passes: boolean}

      const ExploredComponent = explore<ExampleComponentProps, {}>({})(BasicComponent)
      const wrapper = shallow(<ExploredComponent passes={true} />)

      expect(wrapper).toMatchSnapshot()
    })

    it('maps basic props', () => {
      type ExampleComponentProps = {
        passes: boolean,
      }
      type ExploredProps = {
        didThisPass: string,
      }

      const ExploredComponent = explore<ExampleComponentProps, ExploredProps>({
        didThisPass: (props) => props.passes ? 'yes' : 'no',
      })(BasicComponent)

      expect(shallow(<ExploredComponent passes={true} />)).toMatchSnapshot()
    })

    it('maps successful async props', async () => {
      type ExploredProps = {
        kind: string,
      }

      const promise = Promise.resolve('I\'m a promise!')
      const ExploredComponent = explore<{}, ExploredProps>({
        kind: async () => promise,
      })(BasicComponent)
      const wrapper = shallow(<ExploredComponent />)

      await promise
      wrapper.update()
      expect(wrapper).toMatchSnapshot()
    })

    it('maps undefined from errored async props', async () => {
      type ExploredProps = {
        kind: string,
      }

      const promise = Promise.reject('I\'m a failed promise!')
      const ExploredComponent = explore<{}, ExploredProps>({
        kind: async () => promise,
      })(BasicComponent)
      const wrapper = shallow(<ExploredComponent />)

      try {
        await promise
      } catch (e) {
        // ignore
      } finally {
        wrapper.update()
        expect(wrapper).toMatchSnapshot()
      }
    })

    it('maps as a decorator', () => {
      type ExampleComponentProps = {passes: boolean}
      type ExploredProps = {didThisPass?: string}

      /**
       * Example basic decorator class
       */
      @explore<ExampleComponentProps, ExploredProps>({
        didThisPass: (props) => props.passes ? 'yes' : 'no',
      })
      class ExampleComponent extends React.PureComponent<ExampleComponentProps & ExploredProps> {
        public render() {
          const {didThisPass} = this.props

          return <div>Did this pass? {didThisPass}. {didThisPass} it did.</div>
        }
      }

      expect(shallow(<ExampleComponent passes={true} />)).toMatchSnapshot()
    })
  })

  describe('componentWillReceiveProps', () => {
    it('forwards basic props', () => {
      type ExampleComponentProps = {passes: boolean}

      const ExploredComponent = explore<ExampleComponentProps, {}>({})(BasicComponent)
      const wrapper = shallow(<ExploredComponent passes={true} />)
      wrapper.setProps({foo: 'bar'})

      expect(wrapper).toMatchSnapshot()
    })

    it('maps basic props', () => {
      type ExampleComponentProps = {passes:  boolean}
      type ExploredProps = {
        didThisPass:  string,
      }

      const ExploredComponent = explore<ExampleComponentProps,  ExploredProps>({
        didThisPass:  (props) => props.passes ? 'yes' : 'no',
      })(BasicComponent)
      const wrapper = shallow(<ExploredComponent passes={false} />)
      wrapper.setProps({passes:  true})

      expect(wrapper).toMatchSnapshot()
    })

    it('maps async props', async () => {
      type ExploredProps = {
        kind:  string,
      }

      let promise: Promise<string>

      const ExploredComponent = explore<{ foo: boolean }, ExploredProps>({
        kind: async (props) => {
          promise = new Promise((resolve, reject) => {
            if (props.foo) {
              resolve('bar')
            } else {
              resolve('nobar')
            }
          })

          return promise
        },
      })(BasicComponent)
      const wrapper = shallow(<ExploredComponent foo={false}/>)

      await promise
      expect(wrapper).toMatchSnapshot()

      wrapper.setProps({foo:  true})
      await promise
      wrapper.update()
      expect(wrapper).toMatchSnapshot()
    })

    it('maps undefined from errored async props', async () => {
      type ExploredProps = {
        kind:  string,
      }

      const promise = Promise.reject('I\'m a failed promise!')
      const ExploredComponent = explore<{[s:  string]: any }, ExploredProps>({
        kind: async () => promise,
      })(BasicComponent)
      const wrapper = shallow(<ExploredComponent />)

      try {
        await promise
      } catch (e) {
        // ignore
      } finally {
        wrapper.update()
        wrapper.setProps({foo:  true})
        expect(wrapper).toMatchSnapshot()
      }
    })
  })

  // NOTE (JMD): Began work adding a way to handle errors for async props, but put on hold.
  // describe('render with errors', () => {
  //   it('maintains the context', () => {
  //     let storedThis: ExampleComponent
  //
  //     /**
  //      * Example basic decorator class
  //      */
  //     @explore<{}, {}>({})
  //     class ExampleComponent extends React.PureComponent<{}> {
  //       constructor(props: {}) {
  //         super(props)
  //         storedThis = this
  //       }
  //
  //       public render() {
  //         expect(this).toEqual(storedThis)
  //
  //         return <div/>
  //       }
  //     }
  //
  //     expect(shallow(<ExampleComponent/>)).toMatchSnapshot()
  //   })
  //
  //   it('passes props and errors to wrapped render', () => {
  //     type ExampleComponentProps = {passes: boolean}
  //     type ExploredProps = {didThisPass?: string}
  //
  //     /**
  //      * Example basic decorator class
  //      */
  //     @explore<ExampleComponentProps, ExploredProps>({
  //       didThisPass: (props) => props.passes ? 'yes' : 'no',
  //     })
  //     class ExampleComponent extends React.PureComponent<ExampleComponentProps & ExploredProps> {
  //       public render(args?: {props: ExampleComponentProps & ExploredProps}) {
  //         // expect(args).toMatchSnapshot()
  //         if (args !== undefined) {
  //           const {didThisPass} = args.props
  //
  //           return <div>Did this pass? {didThisPass}. {didThisPass} it did.</div>
  //         }
  //
  //         return <div/>
  //       }
  //     }
  //
  //     expect(shallow(<ExampleComponent passes={true}/>)).toMatchSnapshot()
  //   })
  //
  //   it('passes props and errors to wrapped SFC', () => {
  //     type ExampleComponentProps = {passes: boolean}
  //     type ExploredProps = {didThisPass?: string}
  //     type Args = {props?: ExampleComponentProps & ExploredProps & {children?: React.ReactNode}}
  //
  //     const ExampleComponent: React.SFC<ExampleComponentProps & ExploredProps> = (args?: Args) => {
  //       console.log(args)
  //       if (args !== undefined) {
  //         const {didThisPass} = args.props
  //
  //         return <div>Did this pass? {didThisPass}. {didThisPass} it did.</div>
  //       }
  //
  //       return <div/>
  //     }
  //
  //     /**
  //      * Example basic decorator class
  //      */
  //     const WrappedComponent = explore<ExampleComponentProps, ExploredProps>({
  //       didThisPass: (props) => props.passes ? 'yes' : 'no',
  //     })(ExampleComponent)
  //
  //     expect(shallow(<ExampleComponent passes={true}/>)).toMatchSnapshot()
  //   })
  // })
})
