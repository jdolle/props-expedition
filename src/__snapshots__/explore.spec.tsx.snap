// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`explore() componentWillReceiveProps forwards basic props 1`] = `
<BasicComponent
  foo="bar"
  passes={true}
/>
`;

exports[`explore() componentWillReceiveProps maps async props 1`] = `
<BasicComponent
  foo={false}
  kind="nobar"
/>
`;

exports[`explore() componentWillReceiveProps maps async props 2`] = `
<BasicComponent
  foo={true}
  kind="bar"
/>
`;

exports[`explore() componentWillReceiveProps maps basic props 1`] = `
<BasicComponent
  didThisPass="yes"
  passes={true}
/>
`;

exports[`explore() componentWillReceiveProps maps undefined from errored async props 1`] = `
<BasicComponent
  foo={true}
/>
`;

exports[`explore() constructor forwards basic props 1`] = `
<BasicComponent
  passes={true}
/>
`;

exports[`explore() constructor forwards nothing for {} 1`] = `<BasicComponent />`;

exports[`explore() constructor maps as a decorator 1`] = `
<ExampleComponent
  didThisPass="yes"
  passes={true}
/>
`;

exports[`explore() constructor maps basic props 1`] = `
<BasicComponent
  didThisPass="yes"
  passes={true}
/>
`;

exports[`explore() constructor maps successful async props 1`] = `
<BasicComponent
  kind="I'm a promise!"
/>
`;

exports[`explore() constructor maps undefined from errored async props 1`] = `<BasicComponent />`;
